Update spec with data types

thomashoneyman opened the pull request:
This PR updates the `SPEC.md` file with richer descriptions of the various data types used by the registry. Along the way I have added links to the "source code" for these data types, but in practice those files don't yet exist (they're in the `trh/new-codecs` branch). Please forgive the broken links, and I'll move the files to their proper location in a new PR to minimize the review load.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+- `hash`: a [`SRIHash`](#srihash) of the tarball contents of the given version
+- `bytes`: the size of the tarball in bytes
+- `publishedTime`: the time the package was published as an `ISO8601` string
```
While working on the `new-codecs` branch, specifically with datetimes, I realized that the RFC3339String implementation used by the `precise-datetime` library is _not_ actually what the ECMAScript standard uses -- it actually uses a simplified ISO8601String:

https://tc39.es/ecma262/#sec-date-time-string-format

They're basically the same, except that the simple ISO8601 format always has 3 milliseconds places and the RFC3339 format used by `precise-datetime` can have between 1 and 3 places (`.000` vs `.0`, or `.520` vs. `.52`). I want to use the ECMAScript format because under the hood we're using JavaScript `DateTime` values and relying on the `Date` object, which required by the spec to be able to parse the simple ISO8601 format. This also lets us drop a JS dependency on `decimal.js` in the registry library, and the full dependency on `precise-datetime`, which is not a core or contrib library.

So across the board I think it's good to switch to the simple ISO8601 format. The only downside is that we will have to regenerate the registry metadata, but that's a quick tweak.

I'm already using the ISO8601 format throughout the `trh/new-codecs` branch.

f-f replied:
I agreed as well that ISO8601 is a better format overall.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+Each unpublished version of a package records three fields:
+
+- `reason`: a plain text string up to 300 characters long explaining why the version was unpublished
```
For plain text descriptions submitted by users, such as descriptions or reasons in the metadata, I've chosen an arbitrary character limit of 300 characters. Not sure how everyone feels about this. I just don't want it to be unbounded, and it shouldn't be used as a bona-fide README, so some reasonably short character limit makes sense.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
 A section specifying how the registry should relate to other infrastructure and the responsibility of that infrastructure as it relates to the registry.
 
-**4.1 Registry Index**
+**4.1 Package Index**
```
I've opted for "Package Index" instead of "Registry Index" and "Package Storage" instead of "Storage Backend" for consistency's sake. We then get things like the "Package Storage", "Package Index", "Package Sets", "Package Metadata", "Package API", and so on, so terms like "Storage", "Index", etc. can be assumed to mean "Package ..."

thomashoneyman replied:
On second thought, we ended up deciding to use `Metadata Index` instead of `Package Metadata` and `Manifest Index` instead of `Package Index` to be more clear what exactly is in the map.

thomashoneyman commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
 
-**5.1 Register a Package**
-Walkthrough of package registration. Includes the various checks performed to ensure a package is OK.
+### Package Operations
```
I've separated the concept of a "package operation" from a "package set operation", because they're quite different in how we authenticate them, how they affect the registry, how the request is submitted (issues vs. the http api), and so on.

thomashoneyman commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
+
+Summarizes what operations the registry will take after a package is published, transferred, or unpublished. This section _could_ be folded back into the previous section, but I worry that doing so would make that section too long and unfocused.
+
+#### 6.1 Update PAckage Index
```
```suggestion
#### 6.1 Update Package Index
```

JordanMartinez commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
 
-**5.3 Update the Package Set**
-Walkthrough of updating the package sets.
+- `/api/publish`: The `Publish` operation
+- `/api/unpublish`: The `Unpublish` operation
+- `/api/transfer`: The `Transfer` operation
```
Should these have a version associated with them? For example
```
/api/v1/publish
```

thomashoneyman replied:
Yes, it's a good idea to version the API. Thanks!

JordanMartinez commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+Package sets are stored in the `registry` repository under the `package-sets` directory. For example, the first package set was published at https://github.com/purescript/registry/blob/main/package-sets/0.0.1.json. Each package set is stored as a JSON file with the following fields:
+
+1. `version`, which is a string containing the package set [`Version`](#version) for this set.
```
Ah... you clarify this in the 'Releasing a Package Set' section. Perhaps link to that here?

JordanMartinez commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
+
+#### 6.1 Update Package Index
+
+A spec of how the registry attempts to update the registry index.
```
```diff
-A spec of how the registry attempts to update the registry index.
+A spec of how the registry attempts to update the package index.
```

JordanMartinez commented on SPEC.md:
```diff
@@ -180,34 +640,53 @@ Third, we attempt to add the rest of the batch of package versions to the packag
 5. Once there are no more packages to consider in the batch, the new package set is ready.
 
-Fourth, we release the new package set (if we could produce one). Automatic package sets follow the below versioning policy:
+Fourth, we release the new package set (if we could produce one). Automatic package sets follow the versioning policy described in [Section 5.5 (Update the Package Set)](#55-update-the-package-set).
```
This link points to the overall section, but not quite the version policy specifically. This could be updated to point to that link.

JordanMartinez reviewed (commented):
Overall looks good to me! Sorry for taking so long to read through this.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+The final point deserves elaboration. Historically, PureScript packages have been registered in the Bower registry using a `purescript-` prefix so as to avoid naming conflicts with JavaScript packages. However, the prefix is not actually part of the package name, so package managers like Spago refer to the package without the prefix. For example, the Bower-registered name `purescript-prelude` ref ...
+
+To prevent users from accidentally registering their packages with an unnecessary prefix, the registry will not accept packages that begin with `purescript-` unless the package author confirms this is what they want with the Registry Trustees.
```
As far as the registry is concerned it doesn’t matter. As far as encouraging a community convention, I think it’s a nice one and I’d encourage it to continue!

thomashoneyman replied:
I've added a little note to this section.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+{
+  "githubOwner": "purescript",
+  "githubRepo": "prelude"
+}
```
It _sort_ of is including `subdir` -- it's optional, so I've tried to show it both in use or omitted.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+`[keytype] [ssh-public-key] [email]`
+
+Note that the email address does not have to be valid, but it must be the address associated with the private key. A JSON example:
```
Perhaps "usable"?

JordanMartinez replied:
Sounds good to me.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+All packages in the registry must have a license that grants permission for redistribution of the source code. Concretely, the registry requires that all packages use an SPDX license and specify an [SPDX license identifier](https://spdx.dev/ids/). `AND` and `OR` conjunctions are allowed. A `License` is represented as a string, which must be a valid SPDX identifier. For example:
+
+`"MIT OR APACHE-2.0"`
```
`WITH` is supported, but it isn't a conjunction. It's used to modify an identifier by applying an exception, if I remember correctly, like `MIT WITH files-exception`. That is still considered a "license identifier".

JordanMartinez replied:
Ok, thanks for clarifying.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+- `version`: a valid [`Version`](#version)
+- `license`: a valid [`License`](#license)
+- `location`: a valid [`Location`](#location)
+- `owners`: a non-empty array of [`Owner`](#owner) (optional)
```
Yep! I can prepend rather than append it. This would become `owners: (optional) a non-empty...`. Alternately, I could take a note from TypeScript and use `?`:

```
- `owners?`: a non-empty ...
```

but I'm not sure if that would be obvious.

JordanMartinez replied:
What about
```
- `owners` (optional): a non empty...
```

- `owners` (optional): a non empty...

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+- `license`: a valid [`License`](#license)
+- `location`: a valid [`Location`](#location)
+- `owners`: a non-empty array of [`Owner`](#owner) (optional)
+- `description`: a description of your library as a plain text string, not markdown, up to 300 characters (optional)
```
I agree with you. The reason it isn't in there right now is we haven't really discussed how it would be used, and so:

> While Pursuit doesn't support this now, it might be something we'll want for the future.

This sort of thing -- "something we'll want for the future" -- is generally omitted, because we can add it in a forward-compatible way. The reason we use `description` is that packages already use it in their `bower.json` files, even though IIRC Pursuit doesn't use it.

So I agree something like that would be nice, but until we have a good discussion on how we'd like to use it I'm holding off adding it to the type.

JordanMartinez replied:
Ok, sounds good.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+- `location`: a valid [`Location`](#location)
+- `owners`: a non-empty array of [`Owner`](#owner) (optional)
+- `description`: a description of your library as a plain text string, not markdown, up to 300 characters (optional)
+- `files`: a non-empty array of globs, where globs are used to match files outside the `src` directory you want included in your package tarball
```
Yes, that's a good idea. I can move the glob comment and the dependencies comment up.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
 
-**3.7 Package Sets**
-Schemas for the package sets
+Each package file is a JSON Lines file where each line is a package manifest, stored in sorted order ascending by version. This format follows the one documented in the Cargo book: https://doc.rust-lang.org/cargo/reference/registries.html#index-format.
```
Are you referring to the link?

JordanMartinez replied:
Sorry for not being clearer. Yes. Specifically,
> This format follows the one documented in the Cargo book: https://doc.rust-lang.org/cargo/reference/registries.html#index-format.

f-f replied:
I agree. Also I'd like to not see the explicit URL, but it should be a link around `the one documented in the Cargo book`

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+**[Source](./lib/src/Registry/Metadata.purs)**
+**[Spec](./specs/v1/Metadata.dhall)**
+
+All packages in the registry have an associated metadata file, which is located in the `metadata` directory of the `registry` repository under the package name. For example, the metadata for the `aff` package is located at: https://github.com/purescript/registry/blob/main/metadata/aff.json. Metadata files are the source of truth on all published and unpublished versions for a particular package,  ...
```
```suggestion
All packages in the registry have an associated metadata file, which is located in the `metadata` directory of the `registry` repository under the package name. For example, the metadata for the `aff` package is located at: https://github.com/purescript/registry/blob/main/metadata/aff.json. Metadata files are the source of truth on all published and unpublished versions for a particular package, on what their content is, as well as where the package is located. Metadata files are produced by the registry, not by package authors, though they take some information from package manifests.
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+Each published version of a package records three fields:
+
+- `hash`: a [`Sha256`](#Sha256) of the tarball contents of the given version
```
```suggestion
- `hash`: a [`Sha256`](#Sha256) of the compressed archive fetched by the registry for this version
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+Each unpublished version of a package records three fields:
+
+- `reason`: a plain text string up to 300 characters long explaining why the version was unpublished
+- `publishedTime`: the time the package was published as an `ISO8601` string
```
```suggestion
- `publishedTime`: the time the version was published, as an `ISO8601` string
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+- `reason`: a plain text string up to 300 characters long explaining why the version was unpublished
+- `publishedTime`: the time the package was published as an `ISO8601` string
+- `unpublishedTime`: the time the package was unpublished as an `ISO8601` string
```
```suggestion
- `unpublishedTime`: the time the version was unpublished, as an `ISO8601` string
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+The metadata for a given package contains up to four fields:
+
+- `location`: a valid [`Location`](#location) representing where the registry is currently fetching this package's source code
```
Not sure if it's idiomatic, but this sounds better to me:
```suggestion
- `location`: a valid [`Location`](#location) representing where the registry is currently fetching this package's source code from
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
 
-**3.4 Manifest**
-Schema for the manifest format.
+The registry maintains an index of all package manifests in the [`registry-index`](https://github.com/purescript/registry-index) repository. This index makes it convenient for package managers to look up the manifest for a particular package version at any time. The package index maintains the invariant that all dependencies of packages in the index are themselves in the index.
```
```suggestion
The registry maintains an index of all package manifests in the [`registry-index`](https://github.com/purescript/registry-index) repository. This index makes it convenient for package managers to look up the manifest for a particular package version at any time. The package index - just like the Registry itself - maintains the invariant that all dependencies of packages in the index are themselves in the index.
```

f-f commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
 
-**3.5 Metadata**
-Schema for the metadata format.
+All manifests for a given package in the registry are stored in the registry index according to the following rules:
```
```suggestion
All manifests for a given package in the registry are cached in the registry index according to the following rules:
```

We are not explicitly speccing anywhere that the Index is a cache and the Registry is perfectly able to work without it.
The suggestion above is the bare minimum we can do, but probably worth having a paragraph about how it's possible to rebuild the index from just (1) the metadata files and (2) all the tarballs

f-f commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
+First, the registry uses the [`PackageName`](#packagename) listed in the JSON payload to read any existing [`Metadata`](#34-metadata) for the package. To be publishable, one of these two conditions must be true:
 
-Policies followed by the registry — though maybe these are better-suited to be stored as separate specs apart from this document.
+1. The package name is not registered (ie. no metadata exists) AND a location is provided in the payload AND the location is not already in use. In this case, the registry will create new metadata to register this package at the given location.
```
```suggestion
1. The package name is not registered (ie. no metadata exists) AND a location is provided in the payload. In this case, the registry will create new metadata to register this package at the given location.
```

I propose we abandon this restriction, as it's not really necessary.

As an example, if we had an `HttpLocation { url :: String }` alongside the existing `Git` and `GitHub` ones, you could have your own server serving packages for the registry to fetch and use the same `url` for different packages, and still have no risk of confusion if you use different `ref`s for different packages, e.g. `ref: "package1/v1.0.1"` and `ref: "package2/v2.2.0"`

f-f replied:
A more compelling problem is the situation where a package owner wants to rename a package.
Since we can't unregister packages, that is effectively an addition.

If the owner wants to keep the same location, the workflow is more involved than it should:
1. add SSH keys to the current package
2. publish a new version so that the registry picks up the keys
3. issue a transfer to a different location
4. register the new package

If we got away with this restriction they'd have to go only trough step 4.

f-f commented on SPEC.md:
```diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
+Notably, the package version will remain listed in the package sets, rendering those package sets invalid. To minimize the potential disruption caused by unpublishing a package, the following conditions must be met for a package to be unpublished:
+
+1. The package version must have been published (ie. you cannot unpublish a version that does not exist).
+2. The version must have been published within the last 48 hours.
```
I'm afraid we'll need an additional condition for DMCA takedowns.

f-f commented on SPEC.md:
````diff
@@ -68,101 +318,311 @@ This is a good time for diagrams! (cc: @AndrewCondon, @JordanMartinez).Two notes
+
+```jsonc
+{ // Sets the package set compiler version to 0.15.2
+  "compiler": "0.15.2",
````
What happens if I submit an update with a compiler version that is lower than the current one in the sets?

thomashoneyman replied:
It should be rejected. I don't think we should be able to downgrade package sets (ie. publish a `0.15.4` set and then start publishing `0.15.2` packages). Once we're on a new series we're on the new series.

f-f replied:
Are we noting anywhere in the spec that we do not accept updates lowering compiler versions? If not, we should note it here.

thomashoneyman replied:
Yes, above this — here's the link:
https://github.com/purescript/registry-dev/pull/533/files#diff-7426c9e3a694ca6015df5f98637912975f2edea23270203ee89a8bdeed246ee0R567

If that doesn't work, then here's the excerpt:

> A package set update is an object with two keys: `compiler`, an optional field that, if set, will update the compiler version used to compile the package sets (it cannot be downgraded), and `packages`, an object where keys are package names and values are either a version number or `null`. A version number indicates the package should be added to the set or updated to the given version, and `null` indicates the package should be dropped from the package set.

thomashoneyman replied:
Some possible contention in the section directly above that:

> 
Anyone can suggest a package set update to the registry. However, community members can only add or upgrade packages in the package set. To remove a package or downgrade a package version, the update must be submitted by a member of the packaging team or a Registry Trustee.

Should we allow packages to be downgraded? I left it in there in case a package version gets unpublished and there is no more-recent version available (so we'd be forced to downgrade it or drop it from the sets altogether).

thomashoneyman replied:
If these two sections check out then this is good to go.

f-f replied:
This is all good! Thanks for the clarifications.

f-f reviewed (commented):
Thanks for typing all of this down! I have probably missed something, but here's my first batch of notes 😄

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+**Git Repository**
+
+An arbitrary Git repository is represented by a JSON `object` with a required `gitUrl` key and optional `subdir` key. The Git URL must be of the form: `http[s]://host.xz/path/to/repo.git`. The registry uses the `git` CLI tool to fetch your repository. A JSON example:
```
This was my proposal, and then we'd update the parsing to match the spec. In short, we want to support any Git URL that we can use `git clone` on, which means disallowing some forms that we can't use (like SSH).

But I've realized that my restriction is too strong on the extension -- for example, SourceHut doesn't use the .git extension:

`git clone https://git.sr.ht/~rj/digitalocean-image`

So perhaps `http[s]://host.xz/path/to/repo[.git]` is correct.

f-f replied:
Let's have some parsing in here, but we shouldn't be too strict on the form of the URL. E.g. I think we can drop the `.git` suffix. (also we are going to implement special cases for SourceHut and GitHub, so we shouldn't worry too much about the details of their URLs)

thomashoneyman replied:
I've updated this to note the relaxed parsing.

thomashoneyman replied:
I think this covers about everything the registry could fetch, ie. the `git` or `http(s)` protocols, and not the `ssh` or filesystem protocols

```regex
(git|http(s)?)(:(//)?)([\w\.@\:/\-~]+)(\.git)?(/)?
```

You can try on regexr with these tests:

```
# These should be accepted
https://github.com/user/project
https://github.com/user/project.git
http://github.com/user/project.git
https://192.168.101.127/user/project.git
http://192.168.101.127/user/project.git
git://host.xz/path/to/repo.git/
git://host.xz/~user/path/to/repo.git/
http://host.xz/path/to/repo.git/
https://host.xz/path/to/repo.git/

# These ones should not be accepted
git@github.com:user/project.git
git@192.168.101.127:user/project.git
ssh://user@host.xz:port/path/to/repo.git/
ssh://user@host.xz/path/to/repo.git/
ssh://host.xz:port/path/to/repo.git/
ssh://host.xz/path/to/repo.git/
ssh://user@host.xz/path/to/repo.git/
ssh://host.xz/path/to/repo.git/
ssh://user@host.xz/~user/path/to/repo.git/
ssh://host.xz/~user/path/to/repo.git/
ssh://user@host.xz/~/path/to/repo.git
ssh://host.xz/~/path/to/repo.git
/path/to/repo.git/
path/to/repo.git/
~/path/to/repo.git
file:///path/to/repo.git/
file://~/path/to/repo.git/
user@host.xz:/path/to/repo.git/
host.xz:/path/to/repo.git/
user@host.xz:~user/path/to/repo.git/
host.xz:~user/path/to/repo.git/
user@host.xz:path/to/repo.git
host.xz:path/to/repo.git
rsync://host.xz/path/to/repo.git/
```

thomashoneyman commented on SPEC.md:
````diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+}
+```
+
+A note about dependencies: all dependencies you provide must exist in the registry, and the dependency ranges must be solvable (ie. it must be possible to produce a single version of each dependency that satisfies the provided version bounds, including any transitive dependencies).
````
Yep! I think we should ban them, but if we find out we _can't_ do that, because there are too many, then we can relax this.

f-f replied:
Sounds good! I was just remarking it so that we are all aware where the relevant clause is.

thomashoneyman commented on SPEC.md:
```diff
@@ -22,35 +22,285 @@ A simple walkthrough of how a package is published, possibly including diagrams.
+
+A note about dependencies: all dependencies you provide must exist in the registry, and the dependency ranges must be solvable (ie. it must be possible to produce a single version of each dependency that satisfies the provided version bounds, including any transitive dependencies).
+
+A note about globs: globs must contain only `*`, `**`, `/`, `.`, `..`, and characters for Linux file paths. It is not possible to negate a glob (ie. the `!` character), and globs cannot represent a path out of the package source directory.
```
I suppose we could omit it! We always protect against directory traversals with this:

https://github.com/purescript/registry-dev/blob/030a58fe7938e540278c51d551e3ff853d2d3403/lib/src/Registry/Internal/Path.purs#L19

...but I guess we could disallow the use of `..` altogether. Note that right now we just check for directory traversals and then pass off to `fast-glob`, so we don't have explicit parsing for these globs. We'd have to add that later to match the spec.

f-f replied:
I am not sure how much effort it would be to parse globbed paths, but if it's not much effort then it might make sense to do it? I don't have a strong opinion on this, just want to minimise the potential attack surface.

thomashoneyman replied:
It's not hard, with the restricted characters. I can do it in a followup.

thomashoneyman replied:
Well, maybe I spoke too soon. The easiest thing to do is follow the [POSIX portable filename character set](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_276), then add the characters `/` and `*` for globbing.

But this is potentially too restrictive. Alternately, we could ban a selection of characters [such as these](https://en.wikipedia.org/wiki/Filename#Reserved_characters_and_words), as well as common glob characters like `(` or `?`, and ones we definitely don't want, like `..`, and accept any other Unicode.

thomashoneyman commented:
I've addressed all feedback — I think everything is addressed, though these may still need discussion @f-f, @colinwahl:

https://github.com/purescript/registry-dev/pull/533#discussion_r1021971165
https://github.com/purescript/registry-dev/pull/533#discussion_r1021962737
https://github.com/purescript/registry-dev/pull/533#discussion_r1021991346
https://github.com/purescript/registry-dev/pull/533#discussion_r1021975264

Jordan is on PTO and won't be able to continue the review.

thomashoneyman commented:
I believe everything checks out — potentially this still needs discussion:
https://github.com/purescript/registry-dev/pull/533#discussion_r1025834358

As followup tasks, we should:

1. Implement the Git URL parsing and the glob parsing from the spec
2. Enforce that compiler versions can be upgraded (not downgraded) in the package sets, and check that the upgrading logic works (we haven't used it yet).

f-f commented:
Sounds great! I will read through the spec once more before we merge.
